File: clients/fsharp/Model/ServerMessageGame.fs
#nowarn "0058"
namespace AiCup2019.Model
type ServerMessageGame = {
    PlayerView: option<PlayerView>;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this.PlayerView with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
    static member readFrom(reader: System.IO.BinaryReader) = {
        PlayerView = match reader.ReadBoolean() with
            | true ->
                Some(
                    PlayerView.readFrom reader
                    )
            | false -> None
    }

File: clients/csharp/Model/ServerMessageGame.cs
namespace AiCup2019.Model
{
    public struct ServerMessageGame
    {
        public Model.PlayerView? PlayerView { get; set; }
        public ServerMessageGame(Model.PlayerView? playerView)
        {
            this.PlayerView = playerView;
        }
        public static ServerMessageGame ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new ServerMessageGame();
            if (reader.ReadBoolean())
            {
                result.PlayerView = Model.PlayerView.ReadFrom(reader);
            } else
            {
                result.PlayerView = null;
            }
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            if (!PlayerView.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                PlayerView.Value.WriteTo(writer);
            }
        }
    }
}
